src/vcf/fasta.cpp

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2018 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/iostreams/copy.hpp>
+
+#include "bioio/bioio.hpp"
+
+#include "util/file_utils.hpp"
+#include "util/logger.hpp"
+#include "util/stream_utils.hpp"
+#include "util/string_utils.hpp"
+#include "vcf/compression.hpp"
+#include "vcf/fasta.hpp"
+#include "vcf/string_constants.hpp"
+
+std::shared_ptr<std::istream> create_gunzipped_stream(std::shared_ptr<std::istream> stream) {
+  std::shared_ptr<std::stringstream> ss(new std::stringstream());
+  try {
+    boost::iostreams::filtering_istream decompressedStream;
+    ebi::vcf::create_uncompressed_stream(*stream, ebi::vcf::GZ, decompressedStream);
+    boost::iostreams::copy(decompressedStream, *ss);
+  } catch (std::exception const &ex) {
+    BOOST_LOG_TRIVIAL(error) << ex.what();
+  }
+
+  return std::static_pointer_cast<std::istream>(ss);
+}
+
+std::shared_ptr<std::istream> create_decompressed_stream(std::shared_ptr<std::istream> stream) {
+  std::vector<char> line;
+  ebi::vcf::get_magic_num(*stream, line);
+  std::string file_extension = ebi::vcf::get_compression("", line);
+
+  if ( file_extension == ebi::vcf::GZ ) {
+    return create_gunzipped_stream(stream);
+  }
+
+  return stream;
+}
+
+std::shared_ptr<std::istream> create_input_stream(const std::string &path) {
+  if ( !ebi::util::is_remote_url(path) ) {
+    std::shared_ptr<std::ifstream> fstream(new std::ifstream());
+    ebi::util::open_local(*fstream, path, std::ifstream::binary);
+    return create_decompressed_stream(std::static_pointer_cast<std::istream>(fstream));
+  }
+
+  std::shared_ptr<std::stringstream> sstream(new std::stringstream());
```

Comment:
any particular reason for using sstream instead of ifstream ?

Reply:
First it has to be a stream for both input and output because it is getting written to when downloading and re-used laster as an input to assembly checker. Definitly not `ifstream`, `fstream` could do. But I don't want to suggest this stream is associated with a file, it is merely a `string`.

Reply:
ohh my bad. I havent seen implementation of open_remote. I thought it could be like first downloading the file and then opening.
But I think downloading and then opening seems good idea in case the file is too large. why to load whole of it into the memory.

Reply:
The new implementation is now based on storing the downloaded sequence in a tmp file and delete the tmp file once it is done with.